pkg/node/address_linux.go

```diff
@@ -57,16 +61,19 @@ retryScope:
 				continue
 			}
 			if len(a.IP) >= ipLen {
-				if ip.IsPublicAddr(a.IP) {
-					ipsPublic = append(ipsPublic, a.IP)
-				} else {
-					ipsPrivate = append(ipsPrivate, a.IP)
-				}
 				// If the IP is the same as the preferredIP, that
 				// means that maybe it is restored from node_config.h,
 				// so if it is present we prefer this one.
 				if a.IP.Equal(preferredIP) {
 					hasPreferred = true
+				} else if a.Flags&unix.IFA_F_SECONDARY > 0 {
```

Comment:
If we want to filter our secondary addresses, I think we should do it all of the time. Hoisting this to the first or second statement in for loop maybe best.

Reply:
In this case, I think `preferredIP` should take precedence over primary/secondary because it's being explicitly provided as an input.

I don't have a strong preference, but respecting `preferredIP` seems less surprising.

Reply:
Maybe we could refactor the code a bit to make this clearer (note the comment update as well):
```go
// If the IP is the same as the preferredIP, that
// means that maybe it is restored from node_config.h,
// so if it is present we prefer this one, even if it
// is a secondary address.
if a.IP.Equal(preferredIP) {
	hasPreferred = true
	continue
}

if a.Flags&unix.IFA_F_SECONDARY > 0 {
	// Skip secondary addresses
	continue
}
```

Reply:
Won't this be a bit problematic in the case where you may have a preferred IP on an interface, but also a secondary IP which overlaps it? You'd potentially select the secondary IP if it was created first I believe. 

Maybe these are edge cases we aren't concerned with?

Reply:
I might be misunderstanding the scenario you described, but I think the preferred IP would still be selected since it has precedence over the ordering *and* the filtering.

Reply:
What I'm trying to get it is, we must explain this function like this, currently, "We filter secondary IP addresses, unless it is preferred. If two preferred IP addresses exist across interfaces, we return the one created first". 

If we always filter secondary, the outcome is more predictable. "We filter secondary IP addresses. IPs are returned in creation order across interfaces". 

Question tho, given your statement here: https://github.com/cilium/cilium/pull/22866#discussion_r1065007987 why do you propose we filter these secondary addresses? It seems like simply sorting by creation order fixes your issue?

Reply:
Filtering secondary addresses isn't needed, I added it only to be consistent with the kernel's [`inet_select_addr`](https://elixir.bootlin.com/linux/v5.11.8/C/ident/inet_select_addr) function. Maybe it's better to remove it for now? I can't think of a use-case where creation order isn't sufficient for selecting a primary IP.

Reply:
I think being consistent with the kernel is good (as long as it's in a separate commit). That avoids surprises for users when they switch from kube-proxy to KPR for example.

Reply:
I pulled the filtering into a separate commit and added this comment:
> All secondary IPs, except the preferredIP, are filtered out.

IMO, this is clear enough since currently, there's only one preferred IP that's provided as an input.